### ISSUE TITLE
Mock fails when multiple tests are run

### ISSUE THREAD

**Issue by TimNFL** [pester/Pester#1505]
## 1. General summary of the issue
I have stumbled across an issue that I am having trouble even explaining correctly. I had a large (by my standards) pester test file that was getting to long and unwieldy so I began splitting the tests into smaller test files. When it was one file, everything worked correctly as expect. When I run just the `pester` command or `invoke-pester` in the directory with the tests everything in the first test succeeds and the second test file fails with all of the mock commands. To reproduce the issue, I was able to simplify the scripts down to the following 

Example.ps1

```
class TestClass {
    
    [string] TestFunction(){
        return (Get-Item .\foo.txt -ErrorAction Stop).Name
    }
}
```
Example.psd1
```
@{
    RootModule = 'Example.psm1'
    ModuleVersion = '0.0.1'
    GUID = '7bb14d76-a7a0-4b13-bd3d-5c289dd22e2d'
    FunctionsToExport = '*'
    CmdletsToExport = '*'
    VariablesToExport = '*'
    AliasesToExport = '*'
}
```
Example.psm1
```
. C:\Temp\example\Example.ps1

Export-ModuleMember -Function *-*
```
.\Tests\Test1.Tests.ps1
```
$ModuleManifestName = 'Example.psd1'
$ModuleManifestPath = "$PSScriptRoot\..\$ModuleManifestName"

Get-Module Example | Remove-Module -Force
Import-Module "$($ModuleManifestPath)" -Force


It 'Test 1' {
    $true | should be $true
}
```
.\Tests\Test2.Tests.ps1
```
$ModuleManifestName = 'Example.psd1'
$ModuleManifestPath = "$PSScriptRoot\..\$ModuleManifestName"

Get-Module Example | Remove-Module -Force
Import-Module "$($ModuleManifestPath)" -Force

InModuleScope -ModuleName Example {
    Mock Get-Item -Verifiable -MockWith {return "hello"}
    It 'Test 2' {
        $Instance = [TestClass]::New()
        {$Instance.TestFunction()} | should not throw
    }
}
```
When I run just `pester` or `invoke-pester`, the first succeeds the second file fails. If I run `pester .\tests\Test2.Tests.ps1` the test succeeds as expected

## 2. Describe Your Environment

Pester version     : 4.10.1 C:\Program Files\WindowsPowerShell\Modules\Pester\4.10.1\Pester.psd1
PowerShell version : 5.1.18362.752
OS version         : Microsoft Windows NT 10.0.18363.0

## 3. Expected Behavior

My expectation is that the tests should run successfully regardless of order if they run successfully individually

## 4.Current Behavior

```
PS C:\temp\example> pester
Executing all tests in '.'

Executing script C:\temp\example\tests\Test1.Tests.ps1
  [+] Test 1 365ms

Executing script C:\temp\example\tests\Test2.Tests.ps1
  [-] Test 2 117ms
    Expected no exception to be thrown, but an exception "Cannot find path 'C:\temp\example\foo.txt' because it does not exist." was thrown from C:\Temp\example\Example.ps1:5 
char:17
        +         return (Get-Item .\foo.txt -ErrorAction Stop).Name
        +                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~.
    11:         {$Instance.TestFunction()} | should not throw
    at <ScriptBlock>, C:\temp\example\tests\Test2.Tests.ps1: line 11
Tests completed in 1.9s
Tests Passed: 1, Failed: 1, Skipped: 0, Pending: 0, Inconclusive: 0 
```
![image](https://user-images.githubusercontent.com/2656286/80204046-26586e00-85f6-11ea-9813-4dde92188591.png)
[example.zip](https://github.com/pester/Pester/files/4528384/example.zip)

**Comment by nohwnd** [comment]
@TimNFL Thanks for the issue. I tried to repro but I see both tests pass no matter how I run it. I tried Invoke-Pester, and Invoke-Pester on the first file, and then the second and also the other way around. Each time in new powershell session, and also in existing session.

Attaching your repro as zip (with the one absolute path fixed to be psscriptroot): 
[timnfl.zip](https://github.com/pester/Pester/files/4526836/timnfl.zip)

**Comment by TimNFL** [comment]
@nohwnd I have attached a screenshot showing an new session and the failure. I've also attached a zip with the files modified for absolute paths. The scripts assume the path will be:
C:\Temp\example
|--Example.ps1
|--Example.psd1
|--Example.psm1
|____Tests
       |--Test1.Tests.ps1
       |--Test2.Tests.ps1

The pester command is run from C:\Temp\Example directory

**Comment by nohwnd** [comment]
@TimNFL does it also repro with the one I attached? Or only if you use these paths?

**Comment by TimNFL** [comment]
@nohwnd It fails on the attachment that you provided also

**Comment by nohwnd** [comment]
Interesting, what is the definition of `pester` that is not coming from the Pester module (unless there was a mistake).

**Comment by TimNFL** [comment]
@nohwnd it is a reference to pester.bat that lives in the bin folder for pester. I have run the command invoke-pester with the same results. here is a screenshot of me running the scripts in your zip file

![image](https://user-images.githubusercontent.com/2656286/80280510-891f3780-86d2-11ea-87c5-c3d6640dcbe4.png)

**Comment by TimNFL** [comment]
I'm not sure if it is relevant or not, but both here and in the actual test files, if I comment out the the following line in Test1.Tests.ps1
`Import-Module "$($ModuleManifestPath)" -Force`
The tests complete without any issues

**Comment by TimNFL** [comment]
@nohwnd After doing a bit more research, `pester.bat` is a batch file that starts a new powershell session and then runs the `invoke-pester` command. as a result, running pester over and over again on just the `Test2.Tests.ps1` file will always succeed, but it only succeeds the first time with `invoke-pester`. I've attached some screenshots of the process that I have been using to give a better idea of what I am talking about

This first screenshot shows me starting a new powershell session and running the `pester` command  on the `Test2.Tests.ps1` multiple times with no changes in between, each test succeeds
![image](https://user-images.githubusercontent.com/2656286/80280926-6a6e7000-86d5-11ea-99c5-edca19a3e014.png)

This screenshot is me starting a new powershell session and running `invoke-pester` on the `Test2.Tests.ps1` multiple times with no changes in between, the first test succeeds the subsequent tests fail
![image](https://user-images.githubusercontent.com/2656286/80280939-7ce8a980-86d5-11ea-9272-1e7c3886cd95.png)

I've also found that if I do as in my previous comment and remove the `Import-Module "$($ModuleManifestPath)" -Force` line in the first file that both tests succeeds the first time using `invoke-pester` but fails after the fact. 

I might be wrong but it seems that there is a problem with importing the module multiple times or the use of multiple InModuleScope blocks. Not sure which one at this point. Is this a know issue? I would still expect the test that I have to succeed and not fail like it is.

To test these theories, I tried using powershell 7 since 6 and above are supposed to handle classses and module scopes better and I am unable to reproduce the issue at all using either method

**Comment by nohwnd** [comment]
Okay I was using PowerShell 7, sorry. I think this will have to do something with The order of import will try to have another look in the evening.

**Comment by TimNFL** [comment]
@nohwnd After rereading the documentation a couple of times, it appears to me that I've been making a couple of bad assumptions. I assumed that `pester` was an alias to `invoke-pester`, I've since found out that its a batch file which creates a new instance of powershell each time. With that in mind when I went back and ran everything from a single script using `invoke-pester` only succeeded on the first run, subsequent runs would fail. because it is failing with the mock command, I want to say the issue I've been having is a case of powershell 5.1 classes not playing nice with pester which is a known issue. using the `pester` command is essentially another method of implementing the documented workaround for this issue of 

> Simply run your Pester tests in a fresh session every time; this is simple to do with Start-Job. I have this proxy function in my PowerShell profile to help with that:

What are you thoughts? if you agree, I think the documentation should be updated to also suggest using the built in `pester` command as an easier alternative to `Start-Job`.
